### ISSUE TITLE
(proposal) Make finer-grained account types

### ISSUE THREAD

**Issue by jniles** [third-culture-software/bhima#1607]
We could catch a lot of errors early by knowing if a certain account is supposed to contain debtors or not.  Or if a certain account should have suppliers in it.  We could do this with finer-grained account types, but it would require some infrastructure.

I propose we create a system where certain account types can be marked as `debtor` accounts or `cash` accounts, or a variety of different account types.  This would drastically reduce errors by preventing a patient from being assigned a stock account type for example - we could filter the accounts a debtor group may be attached to make sure that they are debtor accounts.

The advantage of this is that an enterprise must only set up their chart of accounts correctly and they can avoid a large class of user errors.  Furthermore, the errors will be contained to "invoiced the incorrect debtor account" rather than "invoiced the stock account".

There are a couple ways to implement this - either by creating different account types with an "income/ expense/balance" label (e.g. the cash account is always a balance account, the sale account is always an income account).  Or we could make two different account types (more complex from a user perspective).  I don't think either should be considered unless we believe that this proposal is a good idea.

Interesting in @IMA-WorldHealth/bhima-core's feedback.

**Comment by mbayopanda** [comment]
The main thing that we must have in mind is for what purpose (or usage) we have to finer-grained account types. If there are concrete needs in term of modules or reports that need that, i think it is crucial to have that `finer-grained account types`. 

- What modules or reports need that ?

- There is a column in the table `account` named `classe`, this column cannot help to that needs ?

I think it is good to have a finer grained account types but it must answer concretes needs for modules or reports for avoiding waste time.

**Comment by jniles** [comment]
There are two use cases that I can see being useful:
 1. Limiting a dropdown to only `debtor` accounts based on an account type `DEBTORS`.
 2. Making a report on all your bank accounts (but not your caisses)

These are classically `BALANCE` accounts, but you cannot separate them at a finer level without parsing account numbers.

We could use the `classe` column.   I think it used to simply be the larger account types (`Assets`, `Income`, `Expense`, `Debtor/Creditor`, etc...).  That would need to change.

Basically, this would be a middle ground between account tags (proposed in #209) and account types (which are too large for the scenarios proposed above).

**Comment by jniles** [comment]
As we move more and more towards trying to make BHIMA easier to use, this seems more and more necessary.  It could be done in conjunction with https://github.com/IMA-WorldHealth/bhima-2.X/issues/2225.

It seems like you can go as fine-grained as you want with accounts - for example, [this resource](http://www.smallbusinessdoer.com/lessons/ae-cheat-sheet-part-4-the-5-main-types-of-accounts/) lists 5 main account types:
 1. Asset
 2. Liability
 3. Equity
 4. Revenue
 5. Expense

However, [this resource](http://www.accountingunplugged.com/2008/08/31/chart-of-accounts-more-on-accounting-types/) takes it one step further by adding Cost of Goods Sold, and [this resource](https://www.kashoo.com/blog/what-are-the-different-account-types-in-accounting) expands them to include the former as well as Gain and Loss on Foreign Currency Exchange.

Ultimately, I agree with @mbayopanda that we should have our account type definitions motivated by reports.  IF we expand our account types to the five listed above, we can produce a report [that looks like this](http://www.smallbusinessdoer.com/wp-content/uploads/AE_Cheat_Sheet_2-3.png) as our balance sheet.  Separating by Liabilities and Equity has been requested multiple times in person by our clients at Vanga.

**Comment by jniles** [comment]
Looks like we've gotten these all in.
